### ISSUE TITLE
qcrit seems not to be used for some systems

### ISSUE THREAD

**Issue by mathren90** [COSMIC-PopSynth/COSMIC#351]
I've been trying to check common envelope (CE, evol_type = 7) initiation rates varying qcrits assumptions, and I surprisingly still find common envelope events for qcrit = 1000 and even qcrits=1e99 (for all stellar types). My understanding is that setting the values in the qcrit_array of the init file overrides the qcflag set there, right?

An example system that ends up entering a CE despite the mass ratio being smaller than the qcrit I set is the the following (this is a portion of bpp for this system re-run individually):
```  
 mass_1  kstar_1    mass_2  kstar_2         porb  evol_type
1.168112      2.0  1.029156      1.0  3501.957901        2.0
1.167663      3.0  1.029156      1.0  3503.389718        2.0
1.000150      4.0  1.042196      1.0  1193.333840        2.0
0.963863      5.0  1.042476      1.0  1236.133027        2.0
0.910487      5.0  1.050086      1.0  1201.728901        3.0
0.908565      6.0  1.050578      1.0  1199.513815        2.0
0.563864      6.0  1.242301      1.0  1169.443758        4.0
0.557638     11.0  1.243915      1.0  1170.731767        2.0
0.557638     11.0  1.243915      2.0  1170.731767        2.0
0.557638     11.0  1.243474      3.0  1171.304845        2.0
0.561089     11.0  1.112594      4.0  1265.409075        2.0
0.561160     11.0  1.078674      5.0  1317.914628        2.0
0.563369     11.0  1.030395      6.0  1270.121569        2.0
0.569817     11.0  0.962275      6.0  1234.554662        3.0
0.577363      5.0  0.931438      6.0  1221.185173        2.0
0.577363      6.0  0.931438      6.0  1221.182048        2.0
0.595560      6.0  0.906307      6.0  1186.512152        5.0
0.595560      6.0  0.906307      6.0  1186.512152        7.0
0.576605     11.0  0.531396     11.0   446.366330        8.0
0.576605     11.0  0.531396     11.0   446.366332       10.0 
```
 So this is an initially wide binary that experience a case C RLOF which is stable, the donor star becomes a COWD (kstar_1 = 11), and then during the reverse case C RLOF with the kstar_2 = 6 as donor star, the COWD transitions back to kstar_1 = 5 and then evolves into kstar_1 = 6, and once there is a mass transfer between two kstar_* = 6 they got to a common envelope (evol_type=7) even if the mass ratio is M2/M1=1.5 << 1000. 

This is the initial conditions array for the system above (you should be able to reproduce this):
```
     kstar_1  kstar_2  mass1_binary  mass2_binary         porb       ecc  metallicity   binfrac   tphysf   mass0_1   mass0_2  rad1  rad2  lumin1  lumin2  massc1  massc2  radc1  radc2  menv1  menv2  renv1  renv2  ospin1  ospin2  b_0_1  b_0_2  bacc1  bacc2  tacc1  tacc2  epoch1  epoch2  tms1  tms2  bhspin1  bhspin2  tphys  neta  bwind  hewind  alpha1  lambdaf  ceflag  tflag  ifflag  wdflag  pisn  bhflag  nsflag  cekickflag  cemergeflag  cehestarflag  mxns   pts1  pts2  pts3  ecsn  ecsn_mlow  aic  ussn  sigma  sigmadiv  bhsigmafrac  polar_kick_angle   beta   xi  acc2  epsnov  eddfac  gamma  bdecayfac  bconst    ck  windflag  qcflag  eddlimflag      dtp  randomseed  bhspinflag  bhspinmag  rejuv_fac  rejuvflag  htpmb  ST_cr  ST_tide  bin_num  SNkick_1  SNkick_2  phi_1  phi_2  theta_1  theta_2  qcrit_0  qcrit_1  qcrit_2  qcrit_3  qcrit_4  qcrit_5  qcrit_6  qcrit_7  qcrit_8  qcrit_9  qcrit_10  qcrit_11  qcrit_12  qcrit_13  qcrit_14  qcrit_15  fprimc_0  fprimc_1  fprimc_2  fprimc_3  fprimc_4  fprimc_5  fprimc_6  fprimc_7  fprimc_8  fprimc_9  fprimc_10  fprimc_11  fprimc_12  fprimc_13  fprimc_14  fprimc_15
322      1.0      1.0      1.168112      1.029156  3504.493974  0.728539         0.02  0.413086  13700.0  1.168112  1.029156   0.0   0.0     0.0     0.0     0.0     0.0    0.0    0.0    0.0    0.0    0.0    0.0     0.0     0.0    0.0    0.0    0.0    0.0    0.0    0.0     0.0     0.0   0.0   0.0      0.0      0.0    0.0   0.5    0.0     0.5     1.0      0.5       0      1       0       1    -2       1       3           2            0             2   2.5  0.001  0.01  0.02   2.5        1.4    1     0  265.0     -30.0          1.0              90.0  0.125  0.5   1.5   0.001     1.0   -1.0          1   -3000 -1000         3       2           0  13700.0  1871462035           0        0.0        1.0          0      1      1        0      322    -100.0    -100.0 -100.0 -100.0   -100.0   -100.0     1000     1000     1000     1000     1000     1000     1000     1000     1000     1000      1000      1000      1000      1000      1000      1000  0.095238  0.095238  0.095238  0.095238  0.095238  0.095238  0.095238  0.095238  0.095238  0.095238   0.095238   0.095238   0.095238   0.095238   0.095238   0.095238
```
This issue does not only happen with these stellar types. In my COSMIC population run with the attached init file (i just renamed it adding a txt extension so github would accept it) I have 8283 systems which have an entry of bpp with evol_type == 7 (I would expect zero!), with stellar types of the stars at the onset of the CE 6+6 (like in the example above), 4+3, 8+3, 7+8, 6+5, etc...Many times the low kstar comes from an accreting WD that becomes something else, but not necessarily always.

Now, of course these are all quite pathologic cases with qcrit value unrealistically high, but I was trying to see if I could completely kill the common envelope events, which should be possible setting qcrits to a very high value, so the expected behavior would be to have still stable mass transfer in this case.  Or does COSMIC/BSE enter a common envelope nevertheless if both stars overfill their Roche lobes?

[qcrits_1000.ini.txt](https://github.com/COSMIC-PopSynth/COSMIC/files/4334641/qcrits_1000.ini.txt)

**Comment by katiebreivik** [comment]
I've confirmed that the only systems entering a common envelope with qcrit=1000.0 here are systems which first log 'contact' (evol_type==5.0) before the common envelope. This means that the radii of the two stars are in contact with one another. I think that in this case, a common envelope is inevitable regardless of the critical mass ratio for RLO mass transfer. 

We will cover this on a COSMIC telecon to confirm that the core dev team decides if this is ok; then will add a note to the documentation.

@mathren90: I'd be happy to hear your input on whether you think this is a sufficient response!

**Comment by mathren90** [comment]
Thanks for looking into this. Indeed I think it is fine if COSMIC goes to a common envelope routine regardless of qcrit when there is a contact binary, this enforces a supposedly physical behavior (leaving aside the over-contact binaries that might survive without entering a CE, like VFTS352 and the work on overcontact binary of Marchant et al. 2016, or WUMas stars) even when the user tries unphysical parameters. If you plan to clarify the documentation, that is great and would be a sufficient response!

However, in my runs at qcrit=1000 and qcrit=1e99 I also fine a handful (22 and 10, respectively) of systems for which evol_type goes from 3 straight to 7 bypassing 5. All of these systems involve a NS (kstar of one star is 13). Maybe this is a logging issue only? I suppose NS are treated as point masses, so maybe this cannot allow for evol_type=7? The separations at the onset of the CE are much wider than a ~10km NS radius of course.

One  example of such system is the following, with ZAMS parameters:   M1 = 7.806106     M2 = 5.381412  P=2858.942021  e = 0.601408   (and Z=0.02, the other initial parameters exactly the same as in the previous post):

```     
          tphys            mass_1     mass_2    kstar_1  kstar_2  evol_type     sep
50     44.812395  1.390936  11.105002      6.0      1.0       15.0               
50     44.812395  1.242000  11.105002     13.0      1.0        2.0
50     59.094969  1.242003  10.841902     13.0      2.0        2.0
50     59.146407  1.242008  10.838496     13.0      3.0        2.0
50     59.147088  1.242008  10.838398     13.0      3.0        3.0             423.082556  
50     59.147654  1.242019   5.088326     13.0      3.0        7.0                0.749738  
50     59.147654  1.390936   1.242019     15.0     13.0        8.0               0.0
```
(I have omitted copying the separations for earlier phases of the evolution). I'm surprised you don't find any of these...

**Comment by katiebreivik** [comment]
Small number statistics strikes again. I ran for 100,000 binaries but didn't run into anything like this. I'll have a closer look at the example you provided.

**Comment by katiebreivik** [comment]
Ok! So I've found two things:

1 - Another case where you automatically go into a CE is if the radius of one of the binary components is >= than the periastron distance (this happens often with highly eccentric binaries with a NS + post-MS companion)

2 - In the conditional where the critical mass ratios are checked, there is another conditional: 

...`.and.(q(j1).gt.qc.or.radx(j1).le.radc(j1)))`

The first bit is the critical mass ratio, which I've verified is working correctly (qc=1000), but the second has radx=max(roche radius, core radius) = radc (core radius). So the core radius is >= roche radius. I think this is being hit in Mathieu's example above, but is not being recorded in the bpp? 

Perhaps we want to split out this conditional so that we aren't assuming qc is not being treated properly.
